ssh/src/extension.ts

```diff
@@ -46,41 +53,226 @@ export function activate(context: vscode.ExtensionContext): void {
+							destinationOptions.host = hostAndPort[0] === '%h' ? destinationOptions.host : hostAndPort[0];
+							destinationOptions.port = hostAndPort[1] === '%p' ? destinationOptions.port : Number(hostAndPort[1]);
+
+							// Look for a defined jumpHost
+							const jumpHostBlock = config.find((line:any) => line.param === 'Host' && line.value === jumpHostArg);
+							if (jumpHostBlock) {
+								const jumpHostNameNode = jumpHostBlock.config.find((line:any) => line.param === 'HostName');
+								const jumpHost: string = jumpHostNameNode ? jumpHostNameNode.value : options.host;
+								const jumpPortNode = jumpHostBlock.config.find((line:any) => line.param === 'Port');
+								const jumpPort: number = jumpPortNode ? jumpPortNode.value : options.port;
+								const jumpUserNode = jumpHostBlock.config.find((line:any) => line.param === 'User');
+								const jumpUser: string = jumpUserNode ? jumpUserNode.value : options.username;
+								jumpOptions = {
+									username: jumpUser,
+									port: jumpPort,
+									host: jumpHost
+								}
+							} else {
+								// port -p
+								const jumpPortArg = args.indexOf('-p');
+								const jumpPort = jumpPortArg > 0 ? Number(args[jumpPortArg + 1]) : 22;
+
+								// username@host (username optional)
+								const jumpUserAndHost = jumpHostArg.split('@');
+								const jumpUser = jumpUserAndHost.length > 1 ? jumpUserAndHost[0]: options.username;
+								const jumpHost = jumpUserAndHost[jumpUserAndHost.length-1];
+
+								jumpOptions = {
+									username: jumpUser,
+									port: jumpPort,
+									host: jumpHost
+								}
+							}
+						}
+					} else {
+						throw new Error('We currently only support ssh in ProxyCommand.');
+					}
+				}
+			} catch(err) {
+				logger.log(`Failed to process ${sshConfig}:\n\t${err}\n\tAttempting to use extension settings.`);
+			}
+		}
 
-		// Create an interactive shell with X11 forwarding and leave it open
-		// as long as VS Code is running so that we can point VS Code to its
-		// display.
-		conn.on('ready', () => {
-			resolve(createForwardingShell(conn));
-		});
+		if (jumpOptions) {
+			logger.log(`Connecting to ${getHostString(destinationOptions)} via ${getHostString(jumpOptions)}`);
+			jump.connect(getConnectConfig(jumpOptions));
+		} else {
+			logger.log(`Connecting to ${getHostString(destinationOptions)}`);
+			// Define callbacks that apply to all connection strategies
+			conn.connect(getConnectConfig(destinationOptions));
+		}
+	});
+}
 
-		conn.on('error', (err) => {
-			reject(err);
-		});
+function getHostString(options: ConnectOptions): string {
+	return `${options.username}@${options.host} port ${options.port}`;
+}
+
+function getConnectConfig(options: ConnectOptions) {
+	return {
+		...options,
+		debug: isVerboseLoggingEnabled() ? logVerbose : undefined,
+		...getAuthOptions()
+	}
+}
+
+function createForwardingShell(conn: Client, shouldGetDisplay: boolean): Promise<string> {
```

Comment:
In places that call this, it's not really obvious what the second parameter does:
```ts
createForwardingShell(conn, true);
```

Something like this would make it much more clear:
```ts
function createForwardingShell(conn: client, { getDisplay }: { getDisplay: boolean }) { 
    //...
}
createForwardingShell(conn, { getDisplay: true });
```